packages/daimo-api/src/contract/ethIndexer.ts

```diff
@@ -0,0 +1,189 @@
+export class ETHIndexer {
+  private cachedBalances: Map<string, bigint> = new Map();
+  private latestBalance: Map<Address, [bigint, number]> = new Map();
+  private allETHTransfers: ETHTransfer[] = [];
+
+  private latest = 5699999;
+  private isIndexing = false;
+
+  private listeners: ((ethTransfers: ETHTransfer[]) => void)[] = [];
+
+  constructor(
+    private vc: ViemClient,
+    private uc: UniswapClient,
+    private nameReg: NameRegistry
+  ) {}
+
+  // Watches shovel for new blocks, and indexes ETH transfers from them.
+  // For now, ETH transfers are just batched balance changes since
+  // Shovel doesn't support them.
+  async watch(shovel: Watcher) {
+    setInterval(async () => {
+      try {
+        if (this.isIndexing) {
+          console.log(`[ETHINDEXER] skipping tick, already indexing`);
+          return;
+        }
+        this.isIndexing = true;
+
+        const shovelLatest = await shovel.getShovelLatest();
+
+        // localLatest <= 0 when there are no new blocks in shovel
+        // or, for whatever reason, we are ahead of shovel.
+        if (shovelLatest > this.latest) {
+          await this.load(this.latest + 1, shovelLatest);
+          this.latest = shovelLatest;
+        }
+      } finally {
+        this.isIndexing = false;
+      }
+    }, 1000);
+  }
+
+  async batchFetchAndCacheBalances(allAddrs: Address[], blockNum: number) {
+    const batchedAddrs = [...chunks(allAddrs, 1000)];
+
+    for (const batch of batchedAddrs) {
+      console.log(`[ETH] fetching ${batch.length} balances at ${blockNum}`);
+
+      const balances = await (async () => {
+        if (blockNum < chainConfig.offChainUtilsDeployBlock) {
+          return [...Array(batch.length).keys()].map((_) => 0n);
+        } else {
+          return await retryBackoff(`batchGetETHBalances`, () =>
+            this.vc.publicClient.readContract({
+              abi: daimoOffchainUtilsABI,
+              address: daimoOffchainUtilsAddress,
+              functionName: "batchGetETHBalances",
+              args: [batch],
+              blockNumber: BigInt(blockNum),
+            })
+          );
+        }
+      })();
+
+      for (const [i, balance] of balances.entries()) {
+        this.cachedBalances.set(addrBlockNumKey(batch[i], blockNum), balance);
+      }
+    }
+  }
+
+  async load(from: number, to: number) {
+    const startTime = Date.now();
+
+    // TODO: For now, ETH transfers are just batched balance changes between
+    // (from, to] since Shovel doesn't support indexing them.
+
+    const allAddrs = this.nameReg.getAllDAccounts().map((a) => a.addr);
+
+    // Query latest balances and starting balances for all accounts
+    await Promise.all([
+      this.batchFetchAndCacheBalances(allAddrs, to),
+      this.batchFetchAndCacheBalances(allAddrs, from - 1),
+    ]);
+
+    const newTransfers = allAddrs
+      .map((addr) => {
+        const balanceAfter = this.cachedBalances.get(
+          addrBlockNumKey(addr, to)
+        )!;
```

Comment:
cachedBalances will grow by ~3000 entries every block, right? do we ever clear/prune it?

Reply:
possible alternative: delete the cache, do something like this

```ts
const [before, after] = await Promise.all([
  this.batchFetchBalances(allAddrs, to),
  this.batchFetchBalances(allAddrs, from - 1),
]);
  
for (const i in allAddrs) {
  const balanceBefore = before[i];
  const balanceAfter = after[i];
  // ...
}
```

Reply:
yeah I was trying to save the repeated call on previous batch, probably not worth it for code quality.